### PR TITLE
Type detection

### PR DESCRIPTION
**Pull request by joelwurtz**
Replace #29 

WDYT @sagikazarmark ?

### CODE REVIEW

File: tests/fixtures/datetime/expected/Normalizer/TestNormalizer.php [hunk 1]
```diff
@@ -42,6 +42,29 @@ public function denormalize($data, $class, $format = null, array $context = [])
         if (property_exists($data, 'date')) {
             $object->setDate(\DateTime::createFromFormat("Y-m-d\TH:i:sP", $data->{'date'}));
         }
+        if (property_exists($data, 'dateOrNull')) {
+            $value = $data->{'dateOrNull'};
+            if (is_object($data->{'dateOrNull'}) and false !== \DateTime::createFromFormat("Y-m-d\TH:i:sP", $data->{'dateOrNull'})) {
```
Comment (sagikazarmark) [inline tests/fixtures/datetime/expected/Normalizer/TestNormalizer.php]: I wonder why the is_object check is there. This is what I fixed in #30. At this point the data is not an object, but a string.

File: tests/fixtures/datetime/expected/Normalizer/TestNormalizer.php [hunk 2]
```diff
@@ -42,6 +42,29 @@ public function denormalize($data, $class, $format = null, array $context = [])
...
+                $value = \DateTime::createFromFormat("Y-m-d\TH:i:sP", $data->{'dateOrNull'});
+            }
+            if (is_null($data->{'dateOrNull'})) {
+                $value = $data->{'dateOrNull'};
+            }
+            $object->setDateOrNull($value);
+        }
+        if (property_exists($data, 'dateOrNullOrInt')) {
+            $value_1 = $data->{'dateOrNullOrInt'};
+            if (is_object($data->{'dateOrNullOrInt'}) and false !== \DateTime::createFromFormat("Y-m-d\TH:i:sP", $data->{'dateOrNullOrInt'})) {
```
Comment (sagikazarmark) [inline tests/fixtures/datetime/expected/Normalizer/TestNormalizer.php]: Same as above.

File: src/Guesser/ChainGuesser.php [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
 use Joli\Jane\Guesser\Guess\MultipleType;
 use Joli\Jane\Guesser\Guess\Type;
+use Joli\Jane\Guesser\JsonSchema\DateTimeGuesser;
```
Comment (sagikazarmark) [inline src/Guesser/ChainGuesser.php]: Where is this import used?

File: tests/fixtures/datetime/expected/Normalizer/TestNormalizer.php (hunk 1 above)
Comment (sagikazarmark) [inline tests/fixtures/datetime/expected/Normalizer/TestNormalizer.php]: It seems to be a copy paste error :wink:

Comment (joelwurtz) [inline tests/fixtures/datetime/expected/Normalizer/TestNormalizer.php]: Not rebased on this PR IMO, will rebase this on master

Comment (joelwurtz) [inline tests/fixtures/datetime/expected/Normalizer/TestNormalizer.php]: Should be good

**Comment by sagikazarmark** [comment]
LGTM. Might worth to tag a new version as well.

**Pull request merged**